libraries/AP_GPS/AP_GPS_SBP2.cpp

```diff
@@ -0,0 +1,494 @@
+
+    } else if (last_heartbeat.protocol_major < 2) {
+
+        state.status = AP_GPS::NO_GPS;
+        Debug("Received a heartbeat from a SBPv1 device. Current driver does not support this protocol. NO GPS, Dying. ");
+        return false;
+
+    } else if (last_heartbeat.nap_error   == 1 ||
+               last_heartbeat.io_error    == 1 ||
+               last_heartbeat.sys_error   == 1) {
+
+        state.status = AP_GPS::NO_GPS;
+
+        Debug("Piksi reported an error. Printing Health Report. NO GPS, Dying.");
+        Debug("   ext_antenna: %d", last_heartbeat.ext_antenna);
+        Debug("   res2: %d", last_heartbeat.res2);
+        Debug("   protocol_major: %d", last_heartbeat.protocol_major);
+        Debug("   protocol_minor: %d", last_heartbeat.protocol_minor);
+        Debug("   res: %d", last_heartbeat.res);
+        Debug("   nap_error: %d", last_heartbeat.nap_error);
+        Debug("   io_error: %d", last_heartbeat.io_error);
+        Debug("   sys_error: %d", last_heartbeat.sys_error);
+
+        return false;
+
+    } else if (last_pos_llh.tow == last_vel_ned.tow
+            && abs((int32_t) (last_gps_time.tow - last_vel_ned.tow)) < 10000
+            && abs((int32_t) (last_dops.tow - last_vel_ned.tow)) < 60000
+            && last_vel_ned.tow > last_full_update_tow) {
+
+        //We have an aligned VEL and LLH, and a recent DOPS and TIME.
+
+        //
+        // Check Flags for Valid Messages
+        //
+        if (last_gps_time.flags.fix_mode    == 0 ||
+            last_vel_ned.flags.fix_mode     == 0 ||
+            last_pos_llh.flags.fix_mode     == 0 ||
+            last_dops.flags.fix_mode        == 0) {
+
+            Debug("Message Marked as Invalid. NO FIX! Flags: {GPS_TIME: %d, VEL_NED: %d, POS_LLH: %d, DOPS: %d}",
+                   last_gps_time.flags.fix_mode,
+                   last_vel_ned.flags.fix_mode,
+                   last_pos_llh.flags.fix_mode,
+                   last_dops.flags.fix_mode);
+
+            state.status = AP_GPS::NO_FIX;
+            return false;
+        }
+
+        //
+        // Update external time and accuracy state
+        //
+        state.time_week         = last_gps_time.wn;
+        state.time_week_ms      = last_vel_ned.tow;
+        state.hdop              = last_dops.hdop;
+        state.vdop              = last_dops.vdop;
+        state.last_gps_time_ms = now;
+
+        //
+        // Update velocity state
+        //
+        state.velocity[0]       = (float)(last_vel_ned.n / 1000.0);
+        state.velocity[1]       = (float)(last_vel_ned.e / 1000.0);
+        state.velocity[2]       = (float)(last_vel_ned.d / 1000.0);
+
+        float ground_vector_sq = state.velocity[0]*state.velocity[0] + state.velocity[1]*state.velocity[1];
+        state.ground_speed = safe_sqrt(ground_vector_sq);
+
+        state.ground_course = wrap_360(degrees(atan2f(state.velocity[1], state.velocity[0])));
+
+        state.speed_accuracy        = 0; //Not supported as of SBP2.1.6
```

Comment:
@njoubert I suspect that it will be a 3D one, but we didn't really document that well in the header... We'll need to make sure we add a comment once we get an answer.

@tridge @priseborough are you expecting the speed_accuracy to be 2D or 3D?